docs/Toy/toy/dialects/toy.py

```diff
@@ -32,13 +33,27 @@
 from xdsl.utils.exceptions import VerifyException
 from xdsl.utils.hints import isa
 
-from xdsl.traits import Pure
+from xdsl.traits import Pure, OpTrait
 
 TensorTypeF64: TypeAlias = TensorType[Float64Type]
 UnrankedTensorTypeF64: TypeAlias = UnrankedTensorType[Float64Type]
 AnyTensorTypeF64: TypeAlias = TensorTypeF64 | UnrankedTensorTypeF64
 
 
+class ToyShapeInferenceTrait(OpTrait, ABC):
+    """
+    Traits Toy operations should inherit from to infer shape inference based on operands.
+    """
+
+    def verify(self, op: Operation) -> None:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def infer_shape(cls, op: Operation) -> None:
+        raise NotImplemented
```

Comment:
```suggestion
        raise NotImplementedError()
```
`NotImplemented` should be only used for notifying that operators (`<`, `==`, `+`) are not implemented, while `NotImplementedError` notify that this will be implemented in the future.
https://docs.python.org/3.6/library/constants.html#NotImplemented
Naming is hard...

Reply:
Ah, ok, thanks for letting me know! I think there are a few more rogue `NotImplemented` around, would be good to fix them also